#' Correct scRNA-seq celltypes based on specific marker genes
#'
#' Correct the automatic celltype annotation result for some clusters based on specific marker genes
#'
#' @docType methods
#' @name RNACorrectCelltype
#' @rdname RNACorrectCelltype
#'
#' @param RNA Seurat object of clustered scRNA-seq dataset, generated by \code{\link{RNARunSeurat}} function.
#' @param genes Data frame of differential expressed genes, generated by \code{\link{RNARunSeurat}} function.
#' @param celltype Cell-type to be corrected. Users can choose from 'nonimmune', 'CD4T/CD8T', 'CD8Tex', 'Tprolif', 'NK', 'pDC'.
#' If non-immune cells need to be corrected, users should provide their own signatures.
#' @param signatures The \code{signatures} should be a data frame of cell-type signatures, with first column is cell-type, and second column is gene symbol.
#' @param score.cutoff Cutoff for cell-type signature score.
#'
#' @author Dongqing Sun
#'
#' @return A RNA Seurat object with corrected celltypes stored in \code{assign.MAESTRO} of \code{meta.data}.
#'
#' @importFrom Seurat DimPlot
#' @importFrom ggplot2 ggsave
#' @export

RNACorrectCelltype <- function(RNA, genes, celltype = "nonimmune", signatures, score.cutoff = NULL) {
    genes$cluster = as.character(genes$cluster)
    if ("assign.MAESTRO" %in% colnames(RNA@meta.data)) {
        celltype_name = "assign.MAESTRO"
    } else {
        celltype_name = "assign.ident"
    }
    cluster_celltype_df = unique(RNA@meta.data[c("seurat_clusters", celltype_name)])
    colnames(cluster_celltype_df)[2] = "celltype"
    cluster_celltype_df$seurat_clusters = as.character(cluster_celltype_df$seurat_clusters)
    cluster_celltype_df$assign.MAESTRO = cluster_celltype_df$celltype
    
    if (celltype == "nonimmune") {
        celltype_list = c("Others", "Endothelial", "Fibroblasts", "Myofibroblasts")
        strategy = "max"
    }
    if (celltype == "CD8Tex") {
        strategy = "positive"
        celltype_list = c("CD8Tex")
        celltype_fuzzy = "CD8T"
        signatures = data.frame(Celltype = c("CD8Tex", "CD8Tex", "CD8Tex", "CD8Tex", "CD8Tex"), Marker = c("LAG3", "PDCD1", "HAVCR2", "CTLA4", "TIGIT"))
        if (is.null(score.cutoff)) {
            score.cutoff = 0.25
        }
    }
    if (celltype == "Tprolif") {
        strategy = "positive"
        celltype_list = c("Tprolif")
        celltype_fuzzy = "Malignant"
        signatures = data.frame(Celltype = c("T", "T", "T"), Marker = c("CD3D", "CD3E", "CD3G"))
        if (is.null(score.cutoff)) {
            score.cutoff = 0
        }
    }
    if (celltype == "NK") {
        strategy = "negative"
        celltype_list = c("NK")
        celltype_fuzzy = "CD8T"
        signatures = data.frame(Celltype = c("CD8T", "CD8T", "CD8T", "CD8T", "CD8T"), Marker = c("CD3D", "CD3E", "CD3G", "CD8A", "CD8B"))
        if (is.null(score.cutoff)) {
            score.cutoff = 0
        }
    }
    if (celltype == "pDC") {
        strategy = "max"
        celltype_list = c("pDC")
        signatures = data.frame(Celltype = c("Mono/Macro", "Mono/Macro", "DC", "DC", "DC"), Marker = c("CSF1R", "CD68", "CLEC4C", "IRF7", "IL3RA"))
    }
    if (celltype == "CD4T/CD8T") {
        strategy = "mixed"
        celltype_list = c("CD4Tconv", "CD8T", "CD8Tex")
        signatures = data.frame(Celltype = c("CD4Tconv", "CD8T", "CD8T"), Marker = c("CD4", "CD8A", "CD8B"))
    }
    
    cluster_df = cluster_celltype_df[cluster_celltype_df$celltype %in% celltype_list, ]
    if (nrow(cluster_df) == 0){
        warning(paste(celltype_list, " not included!"))
    } else {
        cluster_celltype = cluster_df$celltype
        names(cluster_celltype) = cluster_df$seurat_clusters
        cluster_list = names(cluster_celltype)
        cluster_assign.score = sapply(cluster_list, function(x){
            celltype_score = RNAAnnotateCelltypeCluster(genes = genes, signatures = signatures, cluster = x)
            return(celltype_score)
        })
        cluster_assign.score_names = names(cluster_assign.score)
        if (class(cluster_assign.score) != "matrix"){
            cluster_assign.score = matrix(cluster_assign.score, nrow = 1)
            rownames(cluster_assign.score) = unlist(strsplit(cluster_assign.score_names[1], "\\."))[2]
            colnames(cluster_assign.score) = cluster_list
        }

        if (strategy == "max") {
            cluster_celltype_corrected = apply(cluster_assign.score, 2, function(x){
                celltype_corrected = rownames(cluster_assign.score)[which.max(x)]
                return(celltype_corrected)
            })
            names(cluster_celltype_corrected) = colnames(cluster_assign.score)
        }
        if (strategy == "positive") {
            cluster_celltype_corrected = sapply(cluster_list, function(x){
                if (cluster_assign.score[1, x] > score.cutoff) {
                    celltype_corrected =  cluster_celltype[x]
                } else {
                    celltype_corrected = celltype_fuzzy
                }
                return(celltype_corrected)
            })
            names(cluster_celltype_corrected) = cluster_list
        }
        if (strategy == "negative") {
            cluster_celltype_corrected = sapply(cluster_list, function(x){
                if (cluster_assign.score[1, x] < score.cutoff) {
                    celltype_corrected =  cluster_celltype[x]
                } else {
                    celltype_corrected = celltype_fuzzy
                }
                return(celltype_corrected)
            })
            names(cluster_celltype_corrected) = cluster_list
        }
        if (strategy == "mixed") {
            cluster_celltype_corrected = sapply(cluster_list, function(x){
                if (cluster_celltype[x] == "CD4Tconv") {
                    if (cluster_assign.score["CD8T",x] > cluster_assign.score["CD4Tconv",x] & cluster_assign.score["CD8T",x] > 0) {
                        celltype_corrected = "CD8T"
                    } else {
                        celltype_corrected = cluster_celltype[x]
                    }
                } else{
                    if (cluster_assign.score["CD4Tconv",x] > cluster_assign.score["CD8T",x] & cluster_assign.score["CD4Tconv",x] > 0) {
                        celltype_corrected = "CD4Tconv"
                    } else {
                        celltype_corrected = cluster_celltype[x]
                    }
                }
                return(celltype_corrected)
            })
            names(cluster_celltype_corrected) = cluster_list
        }
        
        for (i in names(cluster_celltype_corrected)){
            cluster_celltype_df[cluster_celltype_df$seurat_clusters == i, "assign.MAESTRO"] = cluster_celltype_corrected[i]
        }
        RNA@meta.data$assign.MAESTRO = as.character(RNA@meta.data$seurat_clusters)
        RNA@meta.data$assign.MAESTRO = plyr::mapvalues(x = RNA@meta.data$assign.MAESTRO,
                                                       from = cluster_celltype_df$seurat_clusters, to = cluster_celltype_df$assign.MAESTRO)
    }
    
    p = DimPlot(object = RNA, label = TRUE, pt.size = 0.2, group.by = "assign.MAESTRO", label.size = 3, repel = T)
    ggsave(paste0(RNA@project.name, "_annotated_MAESTRO.png"), p, width=6, height=4)
    return(RNA)
}
